storage/operation/functor.go

```diff
@@ -0,0 +1,281 @@
+// Package operation provides functional programming utilities for database operations
+// with lock context support. It defines functors that can be composed to create
+// complex database operations while ensuring proper lock acquisition and error handling.
+package operation
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+
+	"github.com/jordanschalm/lockctx"
+	"github.com/vmihailenco/msgpack/v4"
+
+	"github.com/onflow/flow-go/module/irrecoverable"
+	"github.com/onflow/flow-go/storage"
+	"github.com/onflow/flow-go/utils/merr"
+)
+
+// Functor represents a database operation that requires a lock context proof
+// and a batch writer. It encapsulates the pattern of acquiring locks before
+// performing database operations and ensures proper error handling.
+type Functor func(lockctx.Proof, storage.ReaderBatchWriter) error
+
+// BindFunctors composes multiple functors into a single functor that executes
+// them sequentially. If any functor fails, the execution stops and returns the error.
+// This enables functional composition of database operations.
+func BindFunctors(functors ...Functor) Functor {
+	return func(lctx lockctx.Proof, rw storage.ReaderBatchWriter) error {
+		for _, fn := range functors {
+			err := fn(lctx, rw)
+			if err != nil {
+				return err
+			}
+		}
+		return nil
+	}
+}
+
+// HoldingLock creates a functor that validates the lock context holds the specified lock.
+// This is used as a guard to ensure operations are only performed when the required lock is held.
+// Returns an error if the lock is not held, otherwise returns nil.
+func HoldingLock(lockID string) Functor {
+	return func(lctx lockctx.Proof, rw storage.ReaderBatchWriter) error {
+		if !lctx.HoldsLock(lockID) {
+			return fmt.Errorf("missing required lock: %s", lockID)
+		}
+		return nil
+	}
+}
+
+// WrapError creates a functor that wraps any error returned by the provided functor
+// with additional context. This is useful for providing more descriptive error messages
+// when composing complex operations.
+func WrapError(wrapMsg string, fn Functor) Functor {
+	return func(lctx lockctx.Proof, rw storage.ReaderBatchWriter) error {
+		err := fn(lctx, rw)
+		if err != nil {
+			return fmt.Errorf("%s: %w", wrapMsg, err)
+		}
+		return nil
+	}
+}
+
+// Overwriting returns a functor that overwrites a key-value pair in the storage.
+// The value is serialized using msgpack encoding. If the key already exists,
+// the value will be overwritten without any checks.
+//
+// This is typically used for operations where we want to update existing data
+// or where we don't care about potential conflicts.
+func Overwriting(key []byte, val interface{}) Functor {
```

Comment:
```suggestion
// Upserting returns a functor that overwrites a key-value pair in the storage.
// The value is serialized using msgpack encoding. If the key already exists,
// the value will be overwritten without any checks.
//
// This is typically used for operations where we want to update existing data
// or where we don't care about potential conflicts.
func Upserting(key []byte, val interface{}) Functor {
```
I prefer the term "upsert": it matches the non-functor function with the same semantics, and better captures that we are _either_ inserting or updating (we are only overwriting if something has already been written).